portfolio/src/main/java/com/google/sps/servlets/PageVisitsServlet.java

```diff
@@ -0,0 +1,98 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.time.LocalDate;
+import java.time.ZoneOffset;
+import java.util.Calendar;
+import java.util.List;
+
+@WebServlet("/visit")
+public class PageVisitsServlet extends HttpServlet {
+
+  private static final String VISITS_TODAY = "Visits Today";
+
+  private static final String VISITS = "visits";
+  private static final String DATE = "date";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    incrementPageVisit(datastore);
+  }
+
+  private void incrementPageVisit(DatastoreService datastore) {
+    LocalDate today = LocalDate.now(ZoneOffset.UTC);
```

Comment:
It's not really about using atomics in Java. The method that does the update will have its own local variables and everything.

The thing that needs to be atomic is "read from DB, increment by one, write to DB".

I don't think "atomic increment" is available out of the box with AppEngine data store (at least, not after a quick search for it). Other systems do have that (e.g. I think MongoDB has an increment operation you can write to the DB).

In Appengine data store, you'd have to use a Transaction to make it happen.

This is why I'm saying don't worry about it for the starter project; blocking a page load on a write transaction is kind of a bad thing to do. You can do it if you want, and if it's easy, because transactions might be important in the Capstone, but don't lose sleep over it. I'd rather you focus on other work (e.g. early start on week 5 stuff). In the Capstone design doc, we'll make sure anything that's transactional is called out.

Reply:
Oh! Sorry. I didn't realize that we were talking about the Datastore in this case. Your example (understandably, given the data) only had integers, so I saw it as a good place to use an AtomicInteger. Needless to say, however, this quite a bit complex. I'll keep situations like this in mind. I know you're saying don't worry about it, and I won't, but I at least want to keep it festering in the back of my mind in case something like this comes up on the capstone and we actually need a solution for it.

In my case, I don't think using a Transaction would be harmful because, while that same data is being polled for creating the chart, it's not necessarily important that the read happens before the write. If I understand how JS promises work, then the chart should be able to fetch data from the servlet while the Transaction is still happening. In that case, couldn't we keep cached data somewhere that is updated, say, every hour in a Transaction of its own?

I see, however, where this could become a much more major issue, given Katie's username situation where users needs to use the data directly where they have it.